src/ucs/arch/x86_64/cpu.c

```diff
@@ -734,16 +734,117 @@ ucs_status_t ucs_arch_get_cache_size(size_t *cache_sizes)
 }
 
 #ifdef __AVX__
+static size_t ucs_x86_nt_all_buffer_transfer(void *dst, const void *src, size_t len)
+{
+    size_t offset;
+    __m256i y0, y1, y2, y3, y4, y5, y6, y7;
+
+    /* copy 64 bytes unconditionally */
+    y6 = _mm256_loadu_si256(src);
+    y7 = _mm256_loadu_si256(UCS_PTR_BYTE_OFFSET(src, 32));
+    _mm256_storeu_si256(dst, y6);
+    _mm256_storeu_si256(UCS_PTR_BYTE_OFFSET(dst, 32), y7);
+
+    offset = 64 - ((uintptr_t)dst & 0x1f);
+    len   -= offset;
+
+    if (ucs_likely((size_t)UCS_PTR_BYTE_OFFSET(src, offset) & 0x1f)) {
+        /* src address is not aligned to 32 byte */
+        while (len >= 256) {
+            y0 = _mm256_loadu_si256(UCS_PTR_BYTE_OFFSET(src, offset));
+            y1 = _mm256_loadu_si256(UCS_PTR_BYTE_OFFSET(src, offset + 32));
+            y2 = _mm256_loadu_si256(UCS_PTR_BYTE_OFFSET(src, offset + 64));
+            y3 = _mm256_loadu_si256(UCS_PTR_BYTE_OFFSET(src, offset + 96));
+            y4 = _mm256_loadu_si256(UCS_PTR_BYTE_OFFSET(src, offset + 128));
+            y5 = _mm256_loadu_si256(UCS_PTR_BYTE_OFFSET(src, offset + 160));
+            y6 = _mm256_loadu_si256(UCS_PTR_BYTE_OFFSET(src, offset + 192));
+            y7 = _mm256_loadu_si256(UCS_PTR_BYTE_OFFSET(src, offset + 224));
+            _mm256_stream_si256(UCS_PTR_BYTE_OFFSET(dst, offset), y0);
+            _mm256_stream_si256(UCS_PTR_BYTE_OFFSET(dst, offset + 32), y1);
+            _mm256_stream_si256(UCS_PTR_BYTE_OFFSET(dst, offset + 64), y2);
+            _mm256_stream_si256(UCS_PTR_BYTE_OFFSET(dst, offset + 96), y3);
+            _mm256_stream_si256(UCS_PTR_BYTE_OFFSET(dst, offset + 128), y4);
+            _mm256_stream_si256(UCS_PTR_BYTE_OFFSET(dst, offset + 160), y5);
+            _mm256_stream_si256(UCS_PTR_BYTE_OFFSET(dst, offset + 192), y6);
+            _mm256_stream_si256(UCS_PTR_BYTE_OFFSET(dst, offset + 224), y7);
+
+            if ((len > 1024) && (((offset >> 8) & 3) == 0)) {
```

Comment:
I understand that when this if() is true, we prefetch range `[src+offset, src+offset+1024[`. Also `offset >> 8` counts iterations (256 bytes at time). So with `... & 3) == 0` we enter the block every 8x256=2048 bytes, to start prefetching not for the next immediate iteration but the one after.

Should not we prefetch 2048 bytes at a time instead of 1024, or prefetch every 1024 bytes instead of 2048, (using 2 instead of 3 like `(offset>>8)&2)`)?

Reply:
It is designed to use prefetch with NTA only on the second half of the cache lines on a block of 1024 bytes(16 cache lines). That is the reason only 8 lines are prefetched using 'NTA'.

Reply:
Thanks, re-reading now I see that we prefetch 512 bytes at a time (8 lines) on a 1024 bytes period.